zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -0,0 +1,688 @@
+#[derive(Clone, Debug, Default)]
+pub struct EccCircuit<F: Field> {
+    /// Maximum number of EcAdd operations supported in one instance of the ECC Circuit.
+    pub max_add_ops: usize,
+    /// Maximum number of scalar multiplication operations supported in one instance of the ECC
+    /// Circuit.
+    pub max_mul_ops: usize,
+    /// Maximum number of pairing operations supported in one instance of the ECC Circuit.
+    pub max_pairing_ops: usize,
+
+    /// EcAdd operations provided as witness data to the ECC circuit.
+    pub add_ops: Vec<EcAddOp>,
+    /// EcMul operations provided as witness data to the ECC circuit.
+    pub mul_ops: Vec<EcMulOp>,
+    /// EcPairing operations provided as witness data to the ECC circuit.
+    pub pairing_ops: Vec<EcPairingOp>,
+
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> EccCircuit<F> {
+    pub(crate) fn assign(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        config: &<Self as SubCircuit<F>>::Config,
+        challenges: &Challenges<Value<F>>,
+    ) -> Result<(), Error> {
+        if self.add_ops.len() > self.max_add_ops
+            || self.mul_ops.len() > self.max_mul_ops
+            || self.pairing_ops.len() > self.max_pairing_ops
+        {
+            error!(
+                "add ops = {}, mul ops = {}, pairing ops = {} > max add ops = {}, max mul ops = {}, max pairing ops = {}",
+                self.add_ops.len(),
+                self.mul_ops.len(),
+                self.pairing_ops.len(),
+                self.max_add_ops,
+                self.max_mul_ops,
+                self.max_pairing_ops,
+            );
+            return Err(Error::Synthesis);
+        }
+
+        // keccak powers of randomness.
+        let keccak_powers = std::iter::successors(Some(Value::known(F::one())), |coeff| {
+            Some(challenges.keccak_input() * coeff)
+        })
+        .take(192)
+        .map(|x| QuantumCell::Witness(x))
+        .collect_vec();
+
+        let ecc_chip = EccChip::<F, FpConfig<F, Fq>>::construct(config.fp_config.clone());
+        let fr_chip =
+            FpConfig::<F, Fr>::construct(config.fp_config.range.clone(), 88, 3, modulus::<Fr>());
+        let pairing_chip = PairingChip::construct(config.fp_config.clone());
+        let fp12_chip =
+            Fp12Chip::<F, FpConfig<F, Fq>, Fq12, 9>::construct(config.fp_config.clone());
+
+        let mut first_pass = SKIP_FIRST_PASS;
+
+        let assigned_ec_ops = layouter.assign_region(
+            || "ecc circuit",
+            |region| {
+                if first_pass {
+                    first_pass = false;
+                    return Ok(EcOpsAssigned::default());
+                }
+
+                let mut ctx = config.fp_config.new_context(region);
+
+                macro_rules! assign_ec_op {
+                    ($op_type:ident, $ops:expr, $n_ops:expr, $assign_fn:ident) => {
+                        $ops.iter()
+                            .filter(|op| !op.is_zero())
```

Comment:
ok i see. op.is_zero means EccCircuit cannot process this op, so EvmCircuit must process this op itself.

Reply:
Yes, exactly. In case of `ecAdd`, it's handled here:
https://github.com/scroll-tech/zkevm-circuits/blob/7588b3192ac4a6521ff949cd12509ce8ddb38a2d/zkevm-circuits/src/evm_circuit/execution/precompiles/ec_add.rs#L75-L108